### PR TITLE
[MM-49989] Pass a context.Context to Client4 methods

### PR DESCRIPTION
**Pull request by hanzei**
#### Summary
By allowing consumers of `model.Client4` to pass a `context.Context`, the API allows developers to easily cancel outgoing requests to the MM API. That also includes the possibility of timing out requests that take too long.

Please note that most of the PR deals with migrating the existing tests. All of them use `context.Background()` as none test for a timeout. Tests that are added in the future might add tests that use a more comprehensive `Context`.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-49989

#### Release Note
```release-note
Pass a context.Context to Client4 methods
```

### CODE REVIEW

**Comment by hanzei** [comment]
@agnivade @BenCookie95 Heads up that I'm the next two weeks. Please merge the PR if and only if you approve the PR without any other comments.

**Comment by wiggin77** [comment]
I'm OoO this week, so reassigned to another reviewer.

**Review by agnivade — COMMENTED** [review]
I admit I am unable to review a huge PR properly. I am interested to learn more about the context on this. We have been using these methods without a context since forever. Did we run into a need to cancel requests recently?

I'd also like to see a test which uses context.WithCancel and verifies that the request indeed gets cancelled.

**Comment by wiggin77** [comment]
I have the same questions as [agnivade](https://github.com/agnivade).  The code looks fine, but I wonder what is driving the change.

**Comment by hanzei** [comment]
@agnivade @wiggin77 The main driver for the change is the ability to cancel requests. One use case is https://github.com/mattermost/mattermost-plugin-apps, where a lot of concurrent requests are sent to the MM server, and cancellation helps with saving resources. There are also a couple of customers/community members who asked for request cancellation. A common use case is to send a lot of requests to MM and cancel them all if one fails. Passing a `context` also allows enforcing timeouts easily, a thing that was also requested by the community and the apps plugin.

**Comment by agnivade** [comment]
> A common use case is to send a lot of requests to MM and cancel them all if one fails.

Thanks for the context Ben. It sounds to me like what you are describing is a case of needing queue-like semantics with retry and backoff mechanisms.

Re: timeout, you can set a global timeout in the HTTP [client](https://pkg.go.dev/net/http#Client) but I agree that in case of request specific timeouts, that doesn't help.

In any case, since 8.0 is delayed, we can still go ahead with this. My only request would be to add a test to verify context cancellation.

**Comment by hanzei** [comment]
I've added a cancellation test via [`57e6154` (#22922)](https://github.com/mattermost/mattermost-server/pull/22922/commits/57e6154133f5f73141b0a7a942fd263722d78126).

File: server/public/model/client4_test.go [hunk 1]
```diff
@@ -105,3 +106,46 @@ func TestClient4SetToken(t *testing.T) {
...
+			t.Fatal("request should not hit the server")
+		}))
+
+		client := NewAPIv4Client(server.URL)
+
+		ctx, cancel := context.WithCancel(context.Background())
+
+		cancel()
+
+		_, resp, err := client.GetMe(ctx, "")
```
Comment (agnivade) [inline server/public/model/client4_test.go:122]: In both tests, can we verify that the error is `context.Canceled`?

Comment (hanzei) [reply server/public/model/client4_test.go:122]: Good point, added in 4cbc745f42.

**Review by agnivade — APPROVED** [review]
Nicely done

**Comment by isacikgoz** [comment]
@hanzei this should be good to go after updating the mmctl client calls.

**Comment by hanzei** [comment]
@wiggin77 No problem, you dealt with more important issues then my PR :heart:

**Comment by hanzei** [comment]
Heads up that after talking with @mgdelacroix I've opted to use `context.TODO()` in all invocations of `model.Client4` methods in the `mmctl` production code. We want to add a flag later that allows users to configure a timeout for their commands. The default behaviour would be to use `context.Background()`.

**Comment by isacikgoz** [comment]
> Heads up that after talking with @mgdelacroix I've opted to use `context.TODO()` in all invocations of `model.Client4` methods in the `mmctl` production code. We want to add a flag later that allows users to configure a timeout for their commands. The default behaviour would be to use `context.Background()`.

TODO is fine but we can still use a default timeout (such as 30 secs) and then let use change it.

**Comment by amyblais** [comment]
@hanzei A quick note that feature complete for v8.0 is on June 14th.

**Comment by hanzei** [comment]
@isacikgoz I'm incident to not add additional functionality to this PR. It's already quite large. I'm happy to file a separate ticket for the default timeout.

**Review by isacikgoz — APPROVED** [review]
Thanks @hanzei that would be perfect, would you mind creating the ticket and assign to our team? This looks good to me.

**Comment by cwarnermm** [comment]
@hanzei - Is there any outstanding documentation effort needed for this PR from your perspective?

**Pull request merged**